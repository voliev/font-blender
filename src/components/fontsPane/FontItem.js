import React, { PureComponent } from "react";
import classNames from "classnames";
import PropTypes from "prop-types";
import { CHAR_RANGE } from "../../constants";
import FontInfo from "./FontInfo";
import FontItemControl from "./FontItemControl";
import "./styles/FontItem.css";

class FontItem extends PureComponent {
  constructor() {
    super();

    this.fontItem = React.createRef();

    this.state = {
      fontItem: undefined
    };

    this.requestFont = this.requestFont.bind(this);
    this.handleFontSelectionDeselection = this.handleFontSelectionDeselection.bind(
      this
    );
    this.handleFontPreviewClick = this.handleFontPreviewClick.bind(this);
  }

  requestFont() {
    this.props.fetchFont(this.props.font.family, CHAR_RANGE);
  }

  handleFontSelectionDeselection(e) {
    const { value } = e.currentTarget;
    !this.props.selected
      ? this.props.addFontToSelected(value)
      : this.props.removeFontFromSelected(value);
  }

  handleFontPreviewClick(e) {
    const { family } = e.currentTarget.dataset;
    this.props.startDisplayShowcaseFontPreview(family);
  }

  componentDidUpdate() {
    if (this.props.isVisible && !this.props.font.requested) {
      this.requestFont();
    }
  }

  componentDidMount() {
    this.setState({ fontItem: this.fontItem.current });
  }

  render() {
    const { font, selected, previewFontFamily } = this.props;

    const componentClasses = classNames({
      FontItem: true,
      selected,
      fetching: font.fetching,
      blank: !font.requested,
      preview: font.family === previewFontFamily
    });

    return (
      <li className={componentClasses} ref={this.fontItem}>
        <FontInfo
          {...font}
          fontPreviewClickHandler={this.handleFontPreviewClick}
        />
        <FontItemControl
          {...font}
          selected={selected}
          fontSelectionDeselectionHandler={this.handleFontSelectionDeselection}
          requestFont={this.requestFont}
        />
      </li>
    );
  }
}

FontItem.propTypes = {
  font: PropTypes.exact({
    family: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    variants: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    subsets: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    fetching: PropTypes.bool.isRequired,
    active: PropTypes.bool.isRequired,
    requested: PropTypes.bool.isRequired,
    prefetched: PropTypes.bool.isRequired,
    fetched: PropTypes.bool.isRequired
  }).isRequired,
  selected: PropTypes.bool.isRequired,
  isVisible: PropTypes.bool,
  previewFontFamily: PropTypes.string,
  fetchFont: PropTypes.func.isRequired,
  addFontToSelected: PropTypes.func.isRequired,
  removeFontFromSelected: PropTypes.func.isRequired,
  startDisplayShowcaseFontPreview: PropTypes.func.isRequired
};

export default FontItem;
