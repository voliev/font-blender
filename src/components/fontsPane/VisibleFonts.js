import React, { Component } from "react";
import classNames from "classnames";
import VisibilitySensor from "react-visibility-sensor";
import PropTypes from "prop-types";
import FontItemContainer from "./FontItemContainer";
import "./styles/VisibleFonts.css";

class VisibleFonts extends Component {
  constructor() {
    super();

    this.state = {
      containment: ""
    };

    this.listRef = React.createRef();
  }

  componentDidMount() {
    this.setState({
      containment: this.listRef.current
    });
  }

  render() {
    const { containment } = this.state;
    const { visibleFonts, adjustBundleSize } = this.props;

    const componentClasses = classNames({
      VisibleFonts: true,
      empty: visibleFonts.length === 0
    });

    return (
      <ul className={componentClasses} ref={this.listRef}>
        {visibleFonts.length > 0 && containment !== "" ? (
          visibleFonts.map((font, idx, src) =>
            !font.requested || idx === src.length - 1 ? (
              <VisibilitySensor
                containment={containment}
                key={font.family}
                onChange={idx === src.length - 1 ? adjustBundleSize : null}
                partialVisibility={true}
                intervalDelay={200}
                scrollCheck={true}
                scrollDelay={1000}
              >
                {({ isVisible }) => (
                  <FontItemContainer font={font} isVisible={isVisible} />
                )}
              </VisibilitySensor>
            ) : (
              <FontItemContainer font={font} key={font.family} />
            )
          )
        ) : (
          <li className="no-matches">No matches found...</li>
        )}
      </ul>
    );
  }
}

VisibleFonts.propTypes = {
  adjustBundleSize: PropTypes.func.isRequired,
  visibleFonts: PropTypes.arrayOf(PropTypes.object.isRequired).isRequired
};

export default VisibleFonts;
