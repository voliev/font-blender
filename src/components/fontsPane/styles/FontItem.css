.FontItem {
  display: grid;
  grid-template-columns: 8rem 8rem 2rem;
  grid-template-rows: minmax(2rem, max-content) 1.5rem 1.5rem;
  grid-template-areas:
    "family family controls"
    "category category controls"
    "charsets styles controls";
  cursor: default;
  opacity: 1;
  -webkit-transition: opacity 0.8s, background-color 0.2s;
  transition: opacity 0.8s, background-color 0.2s;
  padding: var(--space-md) var(--space-md) var(--space-md) var(--space-lg);

  position: relative;
}

.FontItem.selected,
.FontItem:hover {
  transition: background-color 0.2s, color 0.2s;
  background-color: var(--almost-white);
}

.FontItem.selected::after,
.FontItem:hover::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  width: .4rem;
  height: 100%;
  background-color: var(--blue);
  animation: growY .1s ease-out 1 forwards;

  transform-origin: 50% 50%;
}

.FontItem .family {
  align-items: flex-start;
  cursor: pointer;
  display: flex;
  grid-area: family;
  justify-content: space-between;
  margin-top: 0;
}

.FontItem .family .family-name {
  /* TODO: Evaluate margin benefits in this case */
  margin: 0 var(--space-xs) 0 0;
}

.FontItem .family .preview-icon {
  align-items: center;
  align-self: flex-start;
  color: var(--grey);
  display: flex;
  opacity: 0;
  transition: opacity 0.1s;
}

.FontItem.preview .family .preview-icon,
.FontItem .family-name:hover~.preview-icon {
  opacity: 1;
  transition: opacity 0.1s;
}

.FontItem.preview .family .preview-icon {
  color: var(--blue);
}

.FontItem .category {
  align-self: center;
  color: var(--dark-grey);
  grid-area: category;
  justify-self: left;
  font-weight: 700;
}

.FontItem .charsets {
  align-self: center;
  color: var(--dark-grey);
  grid-area: charsets;
  justify-self: left;
}

.FontItem .styles {
  align-self: center;
  color: var(--dark-grey);
  grid-area: styles;
  justify-self: left;
}

.FontItem .number {
  color: var(--dark-grey);
  font-weight: 700;
}

.FontItem.empty {
  opacity: 0.4;
  transition: opacity 0.8s;
}

.FontItem.fetching>*,
.FontItem.blank>* {
  filter: blur(6px);
}
