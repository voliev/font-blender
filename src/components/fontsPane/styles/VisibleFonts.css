.VisibleFonts {
  background-color: var(--white);
  display: flex;
  flex-direction: column;
  grid-area: list;
  height: 100%;
  list-style: none;
  margin-bottom: 0;
  margin-top: 0;
  overflow-y: scroll;
  padding: 0;
  position: relative;
  -webkit-overflow-scrolling: touch;
  width: var(--side-pane-width);
}

.VisibleFonts .no-matches {
  align-items: center;
  display: flex;
  font-weight: 700;
  justify-content: center;
  height: 100%;
  width: 100%;
}

.VisibleFonts .empty {
  align-items: center;
  display: flex;
  font-weight: 700;
  justify-content: center;
}