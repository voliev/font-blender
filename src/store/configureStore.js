import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import fonts from "../reducers/fonts";
import filters from "../reducers/filters";
import selectedFonts from "../reducers/selectedFonts";
import showcase from "../reducers/showcase";
import styles from "../reducers/styles";
import notifications from "../reducers/notifications";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default () => {
  const store = createStore(
    combineReducers({
      fonts,
      filters,
      selectedFonts,
      showcase,
      styles,
      notifications
    }),
    composeEnhancers(applyMiddleware(thunk))
  );

  return store;
};
