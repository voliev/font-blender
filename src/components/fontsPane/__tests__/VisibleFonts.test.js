import React from "react";
import { mount, shallow } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import VisibilitySensor from "react-visibility-sensor";
import VisibleFonts from "../VisibleFonts";
import FontItem from "../FontItem";
import state from "../../../fixtures/state";
import visibleFonts from "../../../fixtures/visibleFonts";

const setup = (
  renderFn = shallow,
  propsOverrides = {},
  provideStore = false
) => {
  const props = {
    visibleFonts,
    adjustBundleSize: jest.fn(),
    ...propsOverrides
  };

  if (provideStore) {
    const mockStore = configureStore();
    const store = mockStore(state);

    return {
      ...props,
      wrapper: renderFn(
        <Provider store={store}>
          <VisibleFonts {...props} />
        </Provider>
      )
    };
  }

  return {
    ...props,
    wrapper: renderFn(<VisibleFonts {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  // NOTE: Snapshot test case is irrelevant since `shallow` doesn't
  //       support ref. Without it the component will be rendered with
  //       empty list. `Mount` supports refs but creates useless
  //       huge snapshot even with one element.
});

it("should not crash if visibleFonts array is empty", () => {
  const { wrapper } = setup(mount, { visibleFonts: [] }, true);

  expect(wrapper).toMatchSnapshot();
});

it("should render corresponding message when the list is empty", () => {
  const { wrapper } = setup(mount, { visibleFonts: [] }, true);

  expect(wrapper.find("li").length).toBe(1);
  expect(wrapper.find("li").hasClass("no-matches")).toBeTruthy();
  expect(wrapper.find("li").text()).toBe("No matches found...");
});

it("should receive corresponding props", () => {
  const { wrapper, adjustBundleSize, visibleFonts } = setup(
    mount,
    undefined,
    true
  );

  expect(typeof wrapper.find(VisibleFonts).prop("adjustBundleSize")).toBe(
    "function"
  );
  expect(wrapper.find(VisibleFonts).prop("adjustBundleSize")).toBe(
    adjustBundleSize
  );
  expect(typeof wrapper.find(VisibleFonts).prop("visibleFonts")).toBe("object");
  expect(
    wrapper
      .find(VisibleFonts)
      .prop("visibleFonts")
      .hasOwnProperty("length")
  ).toBeTruthy();
  expect(wrapper.find(VisibleFonts).prop("visibleFonts")).toEqual(visibleFonts);
});

it("should trigger componentDidMount after initial render", () => {
  const componentDidMount = jest.spyOn(
    VisibleFonts.prototype,
    "componentDidMount"
  );

  setup(mount, undefined, true);

  expect(componentDidMount).toHaveBeenCalled();
});

it("should update state after initial render", () => {
  const { wrapper } = setup(mount, undefined, true);
  const containment = wrapper.find(VisibleFonts).state("containment");

  expect(containment).not.toBe("");
  expect(typeof containment).toBe("object");
  expect(containment.nodeType).toBe(1);
  expect(containment.tagName).toBe("UL");
  expect(containment.constructor).toBe(HTMLUListElement);
});

it("should set corresponding class if the list is empty", () => {
  const { wrapper } = setup(shallow, { visibleFonts: [] });

  expect(wrapper.find("ul").hasClass("empty")).toBeTruthy();
  wrapper.setProps({ visibleFonts });
  expect(wrapper.find("ul").hasClass("empty")).toBeFalsy();
});

it("should wrap corresponding fonts in VisibilitySensor component", () => {
  const { wrapper } = setup(mount, undefined, true);

  const items = wrapper.find(VisibilitySensor);

  expect(items.length).toBeGreaterThan(0);

  items.forEach(node => {
    expect(node.find(FontItem).prop("font").requested).toBeFalsy();
  });
});

it("should pass on corresponding props to VisibilitySensor component", () => {
  const { wrapper, adjustBundleSize } = setup(mount, undefined, true);

  const containment = wrapper.find(VisibleFonts).state("containment");

  const items = wrapper.find(VisibilitySensor);

  items.forEach((node, idx) => {
    expect(node.prop("containment")).toEqual(containment);
    expect(node.prop("scrollCheck")).toBeTruthy();
    expect(node.prop("scrollDelay")).toBe(1000);
    expect(node.prop("intervalDelay")).toBe(200);
    expect(node.prop("partialVisibility")).toBeTruthy();

    // onChange handler should be set only on the last item
    // to extend list length update
    if (idx === items.length - 1) {
      expect(node.prop("onChange")).toBe(adjustBundleSize);
    } else {
      expect(node.prop("onChange")).toBe(null);
    }
  });
});

it("should not wrap requested fonts in VisibilitySensor component", () => {
  const { wrapper } = setup(mount, undefined, true);

  const items = wrapper
    .find(FontItem)
    .filterWhere(node => node.prop("font").requested);

  items.forEach(node => {
    expect(node.find(VisibilitySensor).length).toBe(0);
  });
});
