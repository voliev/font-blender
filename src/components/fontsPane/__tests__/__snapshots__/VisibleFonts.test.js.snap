// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not crash if visibleFonts array is empty 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <VisibleFonts
    adjustBundleSize={[MockFunction]}
    visibleFonts={Array []}
  >
    <ul
      className="VisibleFonts empty"
    >
      <li
        className="no-matches"
      >
        No matches found...
      </li>
    </ul>
  </VisibleFonts>
</Provider>
`;
